crates/ruff_server/src/session/workspace/ruff_settings.rs

```diff
@@ -150,6 +151,19 @@ impl<'a> ConfigurationTransformer for EditorConfigurationTransformer<'a> {
             ..Default::default()
         };
 
+        if let Some(config_file_path) = configuration {
+            match open_configuration_file(&config_file_path, project_root) {
+                // if a custom configuration file was given and the configuration preference is
+                // _not_ filesystem-first, return early with the combined editor configuration.
+                Ok(config_from_file) => {
+                    if configuration_preference != ConfigurationPreference::FilesystemFirst {
+                        return editor_configuration.combine(config_from_file);
+                    }
```

Comment:
Haha yeah, three sources of configuration is kind of a lot. But maybe we can reframe this as two sources, since the `config_from_file` is actually just part of the editor configuration, right...?

Reply:
How about this:
```diff
diff --git a/crates/ruff_server/src/session/workspace/ruff_settings.rs b/crates/ruff_server/src/session/workspace/ruff_settings.rs
index ab8a8447e..766b2aa61 100644
--- a/crates/ruff_server/src/session/workspace/ruff_settings.rs
+++ b/crates/ruff_server/src/session/workspace/ruff_settings.rs
@@ -151,18 +151,18 @@ impl<'a> ConfigurationTransformer for EditorConfigurationTransformer<'a> {
             ..Default::default()
         };
 
-        if let Some(config_file_path) = configuration {
+        // Merge in the editor-specified configuration file, if any.
+        let editor_configuration = if let Some(config_file_path) = configuration {
             match open_configuration_file(&config_file_path, project_root) {
-                // if a custom configuration file was given and the configuration preference is
-                // _not_ filesystem-first, return early with the combined editor configuration.
-                Ok(config_from_file) => {
-                    if configuration_preference != ConfigurationPreference::FilesystemFirst {
-                        return editor_configuration.combine(config_from_file);
-                    }
+                Ok(config_from_file) => editor_configuration.combine(config_from_file),
+                Err(err) => {
+                    tracing::error!("Unable to find custom configuration file {err}");
+                    editor_configuration
                 }
-                Err(err) => tracing::error!("Unable to find custom configuration file {err}"),
             }
-        }
+        } else {
+            editor_configuration
+        };
 
         match configuration_preference {
             ConfigurationPreference::EditorFirst => {
```

Reply:
The basic idea being: the editor configuration is always `editor_configuration.combine(config_from_file)` (if it exists); then, we just follow the match below.

Reply:
Wouldn't this mean that the custom configuration file would get merged with filesystem configuration, though? I thought we wanted to ignore filesystem configuration similar to `--config`.

That being said, we already have another way to ignore filesystem configuration, so maybe those behaviors should be detached and it's up to the user. But it might not be the most intuitive experience.

Reply:
Sorry, I meant to say `EditorOnly` instead of `EditorFirst`. I'm fine with that as a solution, I just wanted to make sure we were on the same page 👍

Reply:
I think it's ok.